packages/grid/x-data-grid-pro/src/utils/tree/utils.ts

```diff
@@ -0,0 +1,235 @@
+          [groupingKey.toString()]: node.id,
+        },
+      },
+      children: [...parentNode.children, node.id],
+    };
+  }
+};
+
+/**
+ * Removes a node from the tree
+ */
+export const removeNodeFromTree = ({
+  node,
+  tree,
+  treeDepths,
+}: {
+  node: GridTreeNode;
+  tree: GridRowTreeConfig;
+  treeDepths: GridTreeDepths;
+}) => {
+  // 1. Remove node from the tree.
+  delete tree[node.id];
+
+  // 2. Decrement the `treeDepths` object for the node's depth.
+  const nodeDepth = node.depth;
+  const currentNodeCount = treeDepths[nodeDepth];
+  if (currentNodeCount === 1) {
+    delete treeDepths[nodeDepth];
+  } else {
+    treeDepths[nodeDepth] = currentNodeCount - 1;
+  }
+
+  // 3. Unregister the new node in its parent.
+  const parentNode = tree[node.parent!] as GridGroupNode;
+  // For footers,
+  // Unregister the node from its parent `footerId` property.
+  if (node.type === 'footer') {
+    tree[parentNode.id] = {
+      ...parentNode,
+      footerId: null,
+    };
+  }
+  // For groups and leaves,
+  // Unregister the node from its parents `children` and `childrenFromPath` properties.
+  else {
+    const groupingField = (node as GridGroupNode).groupingField ?? '__no_field__';
+    const groupingKey = node.groupingKey ?? '__no_key__';
+    const { [groupingKey.toString()]: childrenToRemove, ...newChildrenFromPathWithField } =
+      parentNode.childrenFromPath?.[groupingField] ?? {};
+
+    // TODO rows v6: Can we avoid this linear complexity ?
+    const children = parentNode.children.filter((childId) => childId !== node.id);
+    const childrenFromPath: GridChildrenFromPathLookup = { ...parentNode.childrenFromPath };
+    if (Object.keys(newChildrenFromPathWithField).length === 0) {
+      delete childrenFromPath[groupingField];
+    } else {
+      childrenFromPath[groupingField] = newChildrenFromPathWithField;
+    }
+
+    tree[parentNode.id] = {
+      ...parentNode,
+      children,
+      childrenFromPath,
+    };
+  }
+};
+
+/**
+ * Updates the `id` and `isAutoGenerated` properties of a group node.
+ */
+export const updateGroupNodeIdAndAutoGenerated = ({
+  node,
+  updatedNode,
+  tree,
+  treeDepths,
+}: {
+  node: GridGroupNode;
+  updatedNode: Pick<GridGroupNode, 'id' | 'isAutoGenerated'>;
+  tree: GridRowTreeConfig;
+  treeDepths: GridTreeDepths;
+}) => {
+  // 1. Set the new parent for all children from the old group
+  node.children.forEach((childId) => {
+    tree[childId] = {
+      ...tree[childId],
+      parent: updatedNode.id,
+    };
+  });
+
+  // 2. Remove the old group from the tree
+  removeNodeFromTree({
+    node,
+    tree,
+    treeDepths,
+  });
+
+  // 3. Add the new group in the tree
+  const groupNode: GridGroupNode = {
+    ...node,
+    ...updatedNode,
+  };
+
+  insertNodeInTree({
+    node: groupNode,
+    tree,
+    treeDepths,
+  });
+};
+
+export const addActionToGroup = (
+  updatedGroups: GridRowTreeUpdatedGroups | undefined,
```

Comment:
`updatedGroup` let's think it's a list of groups which have been updated (no idea about why their is an "action" in `addActionToGroup`)

Now that I see what it does, I would rename `updatedGroups` by `actionsAppliedPerGroup` or `groupUpdateApplied`. Something to know this object is per group and per action

Reply:
Since it's now a manager how would you name it ? 

`appliedGroupUpdateManager` ?
`groupActionManager` ?